### PR TITLE
Power gate DSP SRAM banks for D0->D3 flow

### PR DESCRIPTION
**Pull request by lbetlej**
Power gate DSP SRAM banks for D0->D3 flow from locked L1$. No DSP context is
preserved. On D0->D3 Driver reinitializes DSP HW.

Signed-off-by: Lech Betlej <lech.betlej@linux.intel.com>

### CODE REVIEW

File: src/platform/apollolake/include/platform/asm_memory_management.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	movi \ay, HPSRAM_MASK
+	s32i \ay, \ax, 0
+	memw
+	/* since HPSRAM EBB bank #0 might be used as buffer for legacy
+	 * streaming, should not be checked in status
+	 */
+	movi \ax, 0xfffffffe
+	and \ay, \ay, \ax
+	1 :
+		l32i \ax, \az, 0
```
Comment (lgirdwood) [inline src/platform/apollolake/include/platform/asm_memory_management.h]: indentation looks wrong here ?

File: src/platform/apollolake/include/platform/asm_memory_management.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+#define HPSRAM_MASK	((1 << PLATFORM_HPSRAM_EBB_COUNT) - 1)
+#define LPSRAM_MASK	((1 << PLATFORM_LPSRAM_EBB_COUNT) - 1)
+
+	/* powers down entire hpsram. on entry lirerals and code for section
+	 * from where this code is executed needs to be placed in memory which
+	 * is not HPSRAM (in case when this code is located in HPSRAM,
+	 * lock memory in L1$ or L1 SRAM)
+	 */
+	.macro m_adsphal_hpsram_power_down_entire ax, ay, az
```
Comment (lgirdwood) [inline src/platform/apollolake/include/platform/asm_memory_management.h]: should follow naming convention more closely, so cavs_hpsram_power_off ?

File: src/platform/apollolake/include/platform/asm_memory_management.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+/**
+ * \file platform/apollolake/include/platform/asm_memory_management.h
+ * \brief Macros for power gating memory banks specific for Apollolake
+ * \author Lech Betlej <lech.betlej@linux.intel.com>
+ */
+#ifndef ASM_MEMORY_MANAGEMENT_H
+#define ASM_MEMORY_MANAGEMENT_H
+
```
Comment (lgirdwood) [inline src/platform/apollolake/include/platform/asm_memory_management.h]: We need to check here that ASSEMBLY macro is defined (by make i.e. we are not being included by a C file by mistake). A check and compiler warning would be good here.

**Comment by ranj063** [comment]
@lbetlej should this pull request also take care of uncommenting the line:
        //iipc->pm_prepare_D3 = 1;
in ipc_pm_context_save()? 

Or will you submit a separate one for it?

File: src/platform/apollolake/include/platform/asm_memory_management.h (hunk 1 above)
Comment (lbetlej) [reply src/platform/apollolake/include/platform/asm_memory_management.h]: indentation updated per review comment

File: src/platform/apollolake/include/platform/asm_memory_management.h (hunk 2 above)
Comment (lbetlej) [reply src/platform/apollolake/include/platform/asm_memory_management.h]: naming convention updated per review comment

File: src/platform/apollolake/include/platform/asm_memory_management.h (hunk 3 above)
Comment (lbetlej) [reply src/platform/apollolake/include/platform/asm_memory_management.h]: check and warning added per review comment

**Comment by lbetlej** [comment]
@ranj063  I plan to have a separate commit uncommenting //iipc->pm_prepare_D3 = 1; At first D3 entry/exit flow can be tested with driver w/o memory power gating changes.

**Pull request merged**